Upgrade MKL-DNN to DNNL v1.2

pinzhenx opened the pull request:
## Motivation

This PR upgrades MKL-DNN from v0.20 to DNNL v1.2 and resolves #30300.

DNNL (Deep Neural Network Library) is the new brand of MKL-DNN, which improves performance, quality, and usability over the old version.

This PR focuses on the migration of all existing functionalities, including minor fixes, performance improvement and code clean up. It serves as the cornerstone of our future efforts to accommodate new features like OpenCL support, BF16 training, INT8 inference, etc. and to let the Pytorch community derive more benefits from the Intel Architecture.

<br>

## What's included?

Even DNNL has many breaking changes to the API, we managed to absorb most of them in ideep. This PR contains minimalist changes to the integration code in pytorch. Below is a summary of the changes:

<br>

**General:**

1. Replace op-level allocator with global-registered allocator

```
// before
ideep::sum::compute<AllocForMKLDNN>(scales, {x, y}, z);

// after
ideep::sum::compute(scales, {x, y}, z);
```


The allocator is now being registeted at `aten/src/ATen/native/mkldnn/IDeepRegistration.cpp`. Thereafter all tensors derived from the `cpu_engine` (by default) will use the c10 allocator.

```
RegisterEngineAllocator cpu_alloc(
  ideep::engine::cpu_engine(),
  [](size_t size) {
    return c10::GetAllocator(c10::DeviceType::CPU)->raw_allocate(size);
  },
  [](void* p) {
    c10::GetAllocator(c10::DeviceType::CPU)->raw_deallocate(p);
  }
);
```
------

2. Simplify group convolution

We had such a scenario in convolution where ideep tensor shape mismatched aten tensor: when `groups > 1`, DNNL expects weights tensors to be 5-d with an extra group dimension, e.g. `goihw` instead of `oihw` in 2d conv case.

As shown below, a lot of extra checks came with this difference in shape before. Now we've completely hidden this difference in ideep and all tensors are going to align with pytorch's definition. So we could safely remove these checks from both aten and c2 integration code.

```
// aten/src/ATen/native/mkldnn/Conv.cpp

if (w.ndims() == x.ndims() + 1) {	
  AT_ASSERTM(	
      groups > 1,	
      "Only group _mkldnn_conv2d weights could have been reordered to 5d");	
  kernel_size[0] = w.get_dim(0) * w.get_dim(1);	
  std::copy_n(	
      w.get_dims().cbegin() + 2, x.ndims() - 1, kernel_size.begin() + 1);	
} else {	
  std::copy_n(w.get_dims().cbegin(), x.ndims(), kernel_size.begin());	
}
```

------

3. Enable DNNL built-in cache

Previously, we stored DNNL jitted kernels along with intermediate buffers inside ideep using an LRU cache. Now we are switching to the newly added DNNL built-in cache, and **no longer** caching buffers in order to reduce memory footprint.

This change will be mainly reflected in lower memory usage from memory profiling results. On the code side, we removed couple of lines of `op_key_` that depended on the ideep cache before.

------

4. Use 64-bit integer to denote dimensions

We changed the type of `ideep::dims` from `vector<int32_t>` to `vector<int64_t>`. This renders ideep dims no longer compatible with 32-bit dims used by caffe2. So we use something like `{stride_.begin(), stride_.end()}` to cast parameter `stride_` into a int64 vector.




<br>


**Misc changes in each commit:**


**Commit:** change build options

Some build options were slightly changed, mainly to avoid name collisions with other projects that include DNNL as a subproject. In addition, DNNL built-in cache is enabled by option `DNNL_ENABLE_PRIMITIVE_CACHE`.



Old | New
-- | --
WITH_EXAMPLE | MKLDNN_BUILD_EXAMPLES 
WITH_TEST | MKLDNN_BUILD_TESTS 
MKLDNN_THREADING | MKLDNN_CPU_RUNTIME 
MKLDNN_USE_MKL | N/A (not use MKL anymore)

------


**Commit:** aten reintegration

- aten/src/ATen/native/mkldnn/BinaryOps.cpp

    Implement binary ops using new operation `binary` provided by DNNL

- aten/src/ATen/native/mkldnn/Conv.cpp

    Clean up group convolution checks
    Simplify conv backward integration

- aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp
    
    Simplify prepacking convolution weights

- test/test_mkldnn.py

    Fixed an issue in conv2d unit test: it didn't check conv results between mkldnn and aten implementation before. Instead, it compared the mkldnn with mkldnn as the default cpu path will also go into mkldnn. Now we use `torch.backends.mkldnn.flags` to fix this issue

- torch/utils/mkldnn.py

    Prepack weight tensor on module `__init__` to achieve better performance significantly 


------

**Commit:** caffe2 reintegration

- caffe2/ideep/ideep_utils.h

    Clean up unused type definitions

- caffe2/ideep/operators/adam_op.cc & caffe2/ideep/operators/momentum_sgd_op.cc

   Unify tensor initialization with `ideep::tensor::init`. Obsolete `ideep::tensor::reinit`

- caffe2/ideep/operators/conv_op.cc & caffe2/ideep/operators/quantization/int8_conv_op.cc

    Clean up group convolution checks
    Revamp convolution API

- caffe2/ideep/operators/conv_transpose_op.cc

    Clean up group convolution checks
    Clean up deconv workaround code

------

**Commit:** custom allocator

- Register c10 allocator as mentioned above


<br><br>

## Performance

We tested inference on some common models based on user scenarios, and most performance numbers are either better than or on par with DNNL 0.20.

ratio: new / old | Latency (batch=1 4T) | Throughput (batch=64 56T)
-- | -- | --
pytorch resnet18 | 121.4% | 99.7%
pytorch resnet50 | 123.1% | 106.9%
pytorch resnext101_32x8d | 116.3% | 100.1%
pytorch resnext50_32x4d | 141.9% | 104.4%
pytorch mobilenet_v2 | 163.0% | 105.8%
caffe2 alexnet | 303.0% | 99.2%
caffe2 googlenet-v3 | 101.1% | 99.2%
caffe2 inception-v1 | 102.2% | 101.7%
caffe2 mobilenet-v1 | 356.1% | 253.7%
caffe2 resnet101 | 100.4% | 99.8%
caffe2 resnet152 | 99.8% | 99.8%
caffe2 shufflenet | 141.1% | 69.0% †
caffe2 squeezenet | 98.5% | 99.2%
caffe2 vgg16 | 136.8% | 100.6%
caffe2 googlenet-v3 int8 | 100.0% | 100.7%
caffe2 mobilenet-v1 int8 | 779.2% | 943.0%
caffe2 resnet50 int8 | 99.5% | 95.5%


_Configuration:
Platform: Skylake 8180
Latency Test: 4 threads, warmup 30, iteration 500, batch size 1
Throughput Test: 56 threads, warmup 30, iteration 200, batch size 64_

† Shufflenet is one of the few models that require temp buffers during inference. The performance degradation is an expected issue since we no longer cache any buffer in the ideep. As for the solution, we suggest users opt for caching allocator like **jemalloc** as a drop-in replacement for system allocator in such heavy workloads.

pinzhenx committed:
dnnl upgrade: custom allocator

Replace op-level allocator with global-registered allocator. All tensors derived from the `cpu_engine` will use the c10 allocator.

pinzhenx committed:
dnnl upgrade: aten reintegration

1. Implement binary ops using new DNNL operation `binary`
2. Clean up group convolution checks
3. Simplify conv backward integration
4. Simplify prepacking group convolution weights
5. Prepack weight tensor on module `__init__`
6. Fix issue in conv2d unit test

pinzhenx committed:
dnnl upgrade: caffe2 reintegration

1. Clean up unused type definitions
2. Unify tensor initialization with `ideep::tensor::init`. Obsolete `ideep::tensor::reinit`
3. Clean up group convolution checks
4. Revamp convolution API
5. Clean up deconv workaround code

pinzhenx commented:
@Jianhui-Li, @uyongw, @jgong5, @CaoZhongZ

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 6a88bedc:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=6a88bedc17a1038fde1a68e9f29053fb8cec01da).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4302880) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=88685747) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=88685747))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
CONFLICT (add/add): Merge conflict in .circleci/docker/common/install_conda.sh 
Auto-merging .circleci/docker/common/install_conda.sh 
CONFLICT (add/add): Merge conflict in .circleci/docker/build.sh 
Auto-merging .circleci/docker/build.sh 
CONFLICT (add/add): Merge conflict in .circleci/config.yml 
Auto-merging .circleci/config.yml 
CONFLICT (add/add): Merge conflict in .circleci/cimodel/data/pytorch_build_definitions.py 
Auto-merging .circleci/cimodel/data/pytorch_build_definitions.py 
CONFLICT (add/add): Merge conflict in .circleci/cimodel/data/pytorch_build_data.py 
Auto-merging .circleci/cimodel/data/pytorch_build_data.py 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

vadimkantorov commented:
Should then PyTorch use jemalloc by default?

CaoZhongZ commented:
The performance degradation caused by allocator is large allocation eagerly come from and return to kernel. Page clean imposed by kernel has the complexity of a ReLU, linear to tensor size.

We'll have a more detailed tech gist to explain more.

Jianhui-Li commented:
@gottbrath @VitalyFedyunin This is the DNNL upgrade. Our future BFloat16 work depends on this PR. Appreciate if you can review it and give some feedback.

jgong5 commented:
> Should then PyTorch use jemalloc by default?

We had some related discussions from a separate PR FYI: https://github.com/pytorch/pytorch/pull/20567#issuecomment-511621826
So far, preloading jemalloc is the best thing we can do. @dzhulgakov had the concern to make it by default due to packaging concerns: https://github.com/pytorch/pytorch/pull/20567#issuecomment-513599884

gchanan commented:
can you fix the merge conflicts?

Also, there is an MKL-DNN rebrand to DNNL but the updated environment variables (e.g. MKLDNN_BUILD_EXAMPLES) still use the MKLDNN name?

vpirogov commented:
@gchanan,

Old environment variable names and build options continue to work with DNNL v1.2 to maintain backward compatibility. These will be removed in the next major version though.

gchanan commented:
@vpirogov: I'm asking about the _new_ environment variables (see the Old/New table), which have the MKLDNN name.

dzhulgakov commented:
Regarding jemalloc - if we want to use it for entire python process, we'd need to either compile `python` with it or preload the library with LD_PRELOAD. If we care only about tensor memory though, it should be possible to just invoke jemalloc APIs directly and wrap that into c10's allocator. The downside is that we'd have two malloc implementations competing for resources.

Which model is most affected by it? Seems like ShuffleNet? Is there a one line repro?

dzhulgakov commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -66,28 +63,16 @@ ideep::tensor _mkldnn_conv2d(
-  std::vector<int64_t> kernel_size(x.ndims());
-  // mkldnn conv2d weights could have been re-ordered to 5d by
-  // mkldnn_reorder_conv2d_weight
-  if (w.ndims() == x.ndims() + 1) {
```
did we verify that we never serialized 5d weights previously using JIT serialization? It seems ok (because we do custom setstate in mkldnn_utils.py), but I'd double check that JIT serialized models with grouped convs from previous pytorch can be loaded with this PR

bddppq replied:
We **do** have serialized models with 5d weights (e.g. ResNeXt101 which has grouped conv) and https://github.com/fairinternal/ClassyVision uses it.

jgong5 replied:
> We do have serialized models with 5d weights

@dzhulgakov @bddppq: oh, this doesn't sound right. The custom [serialization code](https://github.com/pytorch/pytorch/blob/master/torch/utils/mkldnn.py#L58) does the right thing to save dense tensors but `to_dense` should return a CPU tensor with compatible format reusable by other devices, e.g. 4-d weight instead of 5-d weight. Shall we fix it this time?

dzhulgakov replied:
yes, it wasn't right and got fixed. Unfortunately we still have old serialized models.

can you check that the previous serialized models work? Because we do serialize the weight at 5d - see example https://colab.research.google.com/drive/1Fb0I6vmx-0eXZm507yiR0IN_WJ4BYDaz

However, the only place that weight goes into is mkldnn_reorder_conv2d_weight which apparently used to accept 5d input. We might need to add a patch there. Can you add a unittest verifying that the previous 5d serialized input is handled correctly? Thanks!

pinzhenx replied:
Thanks for the repro. It's very enlightening.

We didn't take that `RecursiveScriptModule` into account indeed, and we noticed that the recursive module `second.pt` cannot be loaded again, as `mkldnn_reorder_conv2d_weight` actually only accept 4d tensors before. Kind of a potential bug though.

We fixed this by safeguarding the `mkldnn_reorder_conv2d_weight` where we load conv weights. It can accept both 4d and 5d weights now, and all conv2d weights will all have a 4d size, so we can further clean up some workaround in `Convolution.cpp`.

@dzhulgakov @bddppq Please take a look at the following commit about this issue.
```
Commit:
dnnl upgrade: support legacy group conv2d
``

bddppq commented on caffe2/ideep/operators/conv_op.cc:
```diff
@@ -36,95 +36,98 @@ class IDEEPConvOp : public IDEEPConvPoolOpBase {
-        grouped ? (filter.get_dim(0) * filter.get_dim(1)) : filter.get_dim(0));
 
     CAFFE_ENFORCE(4 == X.ndims());
-    CAFFE_ENFORCE(4 == filter.ndims() || (grouped && (group_ > 1)));
```
same concern here https://github.com/pytorch/pytorch/pull/32422#discussion_r371382516

pinzhenx replied:
We haven't serialized caffe2 models before, so I don't think it could be a problem as the 5-d weights only exist at runtime.

Jianhui-Li commented:
> can you fix the merge conflicts?
SH team is on Chinese New Year holiday this week and will rebase after back.  Appreciate your review comments. @gchanan @bddppq @dzhulgakov

pinzhenx commented:
Thank you for taking the time to review this PR.

<br>

> @gchanan: I'm asking about the new environment variables (see the Old/New table), which have the MKLDNN name.

You are correct, these env variables and build options now have new names with the prefix `DNNL_`. But as @vpirogov mentioned, old names still continue to work for backward compatibility.

As for the rebranding, we first need to substitute all MKLDNN/MKL-DNN/mkldnn/mkl-dnn with DNNL/dnnl in the repo, and also need to take care of public APIs like `to_mkldnn()` and other stuff like tensor layout `mkldnn`. We hope this PR not to be overwhelmed by tons of renaming and new APIs, so we kept most of the old names to make this upgrade as smooth as possible, except for few of them in `FindMKLDNN.cmake` where renaming has to be made.

<br>

> @dzhulgakov: Which model is most affected by it? Seems like ShuffleNet? Is there a one line repro?
> If we care only about tensor memory though, it should be possible to just invoke jemalloc APIs directly and wrap that into c10's allocator. The downside is that we'd have two malloc implementations competing for resources.

From existing models we have, we only observe the caffe2 shufflenet has been affected in **throughput test**. I'm sorry we cannot provide you with the repro for now, I might have one for you someday later.

Also, we don't think jemalloc is a must for now, I highlighted it only to point out one caveat of this upgrade, that in some heavy workloads throughput performance may degrade, while the scope of the effects is very limited.

Coderx7 commented:
@pinzhenx, @Jianhui-Li Thanks a lot for this, This was really needed. great job. 
By the way, when are we going to have OpenCL integrated as well? 
Does it require significant changes to the codebase in order to use DNNL's GPU acceleration?

gottbrath commented:
Can we summarize what is needed to approve this PR?

dzhulgakov commented on cmake/Modules/FindMKLDNN.cmake:
```diff
@@ -43,9 +43,9 @@ ELSE(MKL_FOUND)
-IF (NOT MKLDNN_THREADING)
-  SET(MKLDNN_THREADING "OMP:COMP" CACHE STRING "")
-ELSEIF (MKLDNN_THREADING STREQUAL "TBB")
+IF (NOT MKLDNN_CPU_RUNTIME)
```
I think there are other places where this flag is references, including docs: https://github.com/pytorch/pytorch/search?q=MKLDNN_THREADING&unscoped_q=MKLDNN_THREADING

You can try to rename it, but given that it's about the same name - why not to keep original name?

pinzhenx replied:
Thanks for pointing it out. Due it's a change inside DNNL to align with a newly added build option called `DNNL_GPU_RUNTIME`, so we have to use the new name. I have filed another patch to fix this.

dzhulgakov reviewed (commented):
Overall seems good, but the two issues mentioned inline need to be resolved

pinzhenx committed:
dnnl upgrade: support legacy group conv2d

- support legacy mkldnn group conv2d
- add test
- remove workaround for mkldnn group conv2d

pinzhenx committed:
dnnl upgrade: change build options

1. enable DNNL built-in cache
2. Change some DNNL build options
3. replace all occurrences of `MKLDNN_THREADING` with `MKLDNN_CPU_RUNTIME`

pinzhenx commented:
Hi @peterjc123, we are encountering a build error on the Windows platform, however, we are unable to repro it on our side with vs2019. May I please ask for your kind assistance with this issue?

And hi @vpirogov, it seems that this issue occurs in the DNNL CMake files. If you have some ideas, please also let me know, thanks!

Thank you in advance for your help.

peterjc123 commented:
> Hi @peterjc123, we are encountering a build error on the Windows platform, however, we are unable to repro it on our side with vs2019. May I please ask for your kind assistance with this issue?
> 
> And hi @vpirogov, it seems that this issue occurs in the DNNL CMake files. If you have some ideas, please also let me know, thanks!
> 
> Thank you in advance for your help.
> 
> _Here's a copy of CI build log:_
> 
> ```
> cl : Command line warning D9025 : overriding '/GR' with '/GR-'
> cl : Command line warning D9025 : overriding '/w' with '/W4'
> test\googletest\include -I..\third_party\protobuf\src -Iwin_tmp\mkl\include /DWIN32 /D_WINDOWS /GR  /w /EHa /bigobj -openmp:experimental /MD /O2 /Ob2 /DNDEBUG /w /EHa /bigobj   -MP -GR- -GF -Zc:inline -Zc:strictStrings -Zc:threadSafeInit- -W4 -GS- -O2 -Oi -std:c++14 /showIncludes /Fothird_party\fbgemm\asmjit\CMakeFiles\asmjit.dir\src\asmjit\x86\x86assembler.cpp.obj /Fdthird_party\fbgemm\asmjit\CMakeFiles\asmjit.dir\ /FS -c ..\third_party\fbgemm\third_party\asmjit\src\asmjit\x86\x86assembler.cpp 
> Microsoft (R) C/C++ Optimizing Compiler Version 19.24.28316 for x64
> Copyright (C) Microsoft Corporation.  All rights reserved.
> 
> cl : Command line warning D9025 : overriding '/GR' with '/GR-'
> cl : Command line warning D9025 : overriding '/w' with '/W4'
> \core\include -IC:\Users\circleci\project\third_party\pybind11\include -I\opt\rocm\hip\include -I\include -IC:\Users\circleci\project\third_party\cub -IC:\Users\circleci\project\build\win_tmp\magma\include -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\include" -DWIN32   /Qpar /openmp -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 /fo third_party\ideep\mkl-dnn\src\CMakeFiles\dnnl.dir\version.rc.res C:\Users\circleci\project\build\third_party\ideep\mkl-dnn\src\version.rc 
> FAILED: third_party/ideep/mkl-dnn/src/CMakeFiles/dnnl.dir/version.rc.res  
> \core\include -IC:\Users\circleci\project\third_party\pybind11\include -I\opt\rocm\hip\include -I\include -IC:\Users\circleci\project\third_party\cub -IC:\Users\circleci\project\build\win_tmp\magma\include -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\include" -DWIN32   /Qpar /openmp -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 /fo third_party\ideep\mkl-dnn\src\CMakeFiles\dnnl.dir\version.rc.res C:\Users\circleci\project\build\third_party\ideep\mkl-dnn\src\version.rc 
> googletest\googletest\include -I..\third_party\protobuf\src -Iwin_tmp\mkl\include /DWIN32 /D_WINDOWS /GR  /w /EHa /bigobj -openmp:experimental /MD /O2 /Ob2 /DNDEBUG /w /EHa /bigobj   -MP -GR- -GF -Zc:inline -Zc:strictStrings -Zc:threadSafeInit- -W4 -GS- -O2 -Oi -std:c++14 /showIncludes /Fothird_party\fbgemm\asmjit\CMakeFiles\asmjit.dir\src\asmjit\x86\x86instdb.cpp.obj /Fdthird_party\fbgemm\asmjit\CMakeFiles\asmjit.dir\ /FS -c ..\third_party\fbgemm\third_party\asmjit\src\asmjit\x86\x86instdb.cpp 
> Microsoft (R) C/C++ Optimizing Compiler Version 19.24.28316 for x64
> Copyright (C) Microsoft Corporation.  All rights reserved.
> 
> cl : Command line warning D9025 : overriding '/GR' with '/GR-'
> cl : Command line warning D9025 : overriding '/w' with '/W4'
> ogletest\googletest\include -I..\third_party\protobuf\src -Iwin_tmp\mkl\include /DWIN32 /D_WINDOWS /GR  /w /EHa /bigobj -openmp:experimental /MD /O2 /Ob2 /DNDEBUG /w /EHa /bigobj   -MP -GR- -GF -Zc:inline -Zc:strictStrings -Zc:threadSafeInit- -W4 -GS- -O2 -Oi -std:c++14 /showIncludes /Fothird_party\fbgemm\asmjit\CMakeFiles\asmjit.dir\src\asmjit\x86\x86logging.cpp.obj /Fdthird_party\fbgemm\asmjit\CMakeFiles\asmjit.dir\ /FS -c ..\third_party\fbgemm\third_party\asmjit\src\asmjit\x86\x86logging.cpp 
> ..\third_party\fbgemm\third_party\asmjit\src\asmjit\x86\x86logging.cpp(677): warning C4189: 'instInfo': local variable is initialized but not referenced
> Microsoft (R) C/C++ Optimizing Compiler Version 19.24.28316 for x64
> ```

It seems to be caused by the combination of Ninja and cmcldeps. As the following log can be seen.
```
[146/3164] "C:/Jenkins/Miniconda3/Library/bin/cmcldeps.exe" RC C:\Users\circleci\project\build\third_party\ideep\mkl-dnn\src\version.rc third_party\ideep\mkl-dnn\src\CMakeFiles\dnnl.dir\version.rc.res.d third_party\ideep\mkl-dnn\src\CMakeFiles\dnnl.dir\version.rc.res "" "C:/Users/circleci/project/build/win_tmp/bin/sccache.exe" C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\rc.exe -DDNNL_DLL -DDNNL_DLL_EXPORTS -DDNNL_ENABLE_CONCURRENT_EXEC -DDNNL_ENABLE_MAX_CPU_ISA -DDNNL_ENABLE_PRIMITIVE_CACHE -DIDEEP_USE_MKL -DMAGMA_V2 -DNOMINMAX -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DTH_BLAS_MKL -DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_DEPRECATE=1 -D_OPENMP_NOFORCE_MANIFEST -D_WIN -D__STDC_CONSTANT_MACROS -D__STDC_LIMIT_MACROS -IC:\Users\circleci\project\third_party\benchmark\include -IC:\Users\circleci\project\build\caffe2\contrib\aten -IC:\Users\circleci\project\third_party\onnx -IC:\Users\circleci\project\build\third_party\onnx -IC:\Users\circleci\project\third_party\foxi -IC:\Users\circleci\project\build\third_party\foxi -IC:\Users\circleci\project\third_party\ideep\mkl-dnn\include -IC:\Users\circleci\project\build\third_party\ideep\mkl-dnn\include -IC:\Users\circleci\project\third_party\ideep\mkl-dnn\src -IC:\Users\circleci\project\third_party\ideep\mkl-dnn\src\common -IC:\Users\circleci\project\third_party\ideep\mkl-dnn\include -IC:\Users\circleci\project\third_party\googletest\googlemock\include -IC:\Users\circleci\project\third_party\googletest\googletest\include -IC:\Users\circleci\project\third_party\protobuf\src -IC:\Users\circleci\project\build\win_tmp\mkl\include -IC:\Users\circleci\project\third_party -IC:\Users\circleci\project\third_party\eigen -IC:\Jenkins\Miniconda3\include -IC:\Jenkins\Miniconda3\Lib\site-packages\numpy\core\include -IC:\Users\circleci\project\third_party\pybind11\include -I\opt\rocm\hip\include -I\include -IC:\Users\circleci\project\third_party\cub -IC:\Users\circleci\project\build\win_tmp\magma\include -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\include" -DWIN32   /Qpar /openmp -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 /fo third_party\ideep\mkl-dnn\src\CMakeFiles\dnnl.dir\version.rc.res C:\Users\circleci\project\build\third_party\ideep\mkl-dnn\src\version.rc
FAILED: third_party/ideep/mkl-dnn/src/CMakeFiles/dnnl.dir/version.rc.res 
"C:/Jenkins/Miniconda3/Library/bin/cmcldeps.exe" RC C:\Users\circleci\project\build\third_party\ideep\mkl-dnn\src\version.rc third_party\ideep\mkl-dnn\src\CMakeFiles\dnnl.dir\version.rc.res.d third_party\ideep\mkl-dnn\src\CMakeFiles\dnnl.dir\version.rc.res "" "C:/Users/circleci/project/build/win_tmp/bin/sccache.exe" C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\rc.exe -DDNNL_DLL -DDNNL_DLL_EXPORTS -DDNNL_ENABLE_CONCURRENT_EXEC -DDNNL_ENABLE_MAX_CPU_ISA -DDNNL_ENABLE_PRIMITIVE_CACHE -DIDEEP_USE_MKL -DMAGMA_V2 -DNOMINMAX -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DTH_BLAS_MKL -DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_DEPRECATE=1 -D_OPENMP_NOFORCE_MANIFEST -D_WIN -D__STDC_CONSTANT_MACROS -D__STDC_LIMIT_MACROS -IC:\Users\circleci\project\third_party\benchmark\include -IC:\Users\circleci\project\build\caffe2\contrib\aten -IC:\Users\circleci\project\third_party\onnx -IC:\Users\circleci\project\build\third_party\onnx -IC:\Users\circleci\project\third_party\foxi -IC:\Users\circleci\project\build\third_party\foxi -IC:\Users\circleci\project\third_party\ideep\mkl-dnn\include -IC:\Users\circleci\project\build\third_party\ideep\mkl-dnn\include -IC:\Users\circleci\project\third_party\ideep\mkl-dnn\src -IC:\Users\circleci\project\third_party\ideep\mkl-dnn\src\common -IC:\Users\circleci\project\third_party\ideep\mkl-dnn\include -IC:\Users\circleci\project\third_party\googletest\googlemock\include -IC:\Users\circleci\project\third_party\googletest\googletest\include -IC:\Users\circleci\project\third_party\protobuf\src -IC:\Users\circleci\project\build\win_tmp\mkl\include -IC:\Users\circleci\project\third_party -IC:\Users\circleci\project\third_party\eigen -IC:\Jenkins\Miniconda3\include -IC:\Jenkins\Miniconda3\Lib\site-packages\numpy\core\include -IC:\Users\circleci\project\third_party\pybind11\include -I\opt\rocm\hip\include -I\include -IC:\Users\circleci\project\third_party\cub -IC:\Users\circleci\project\build\win_tmp\magma\include -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\include" -DWIN32   /Qpar /openmp -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 /fo third_party\ideep\mkl-dnn\src\CMakeFiles\dnnl.dir\version.rc.res C:\Users\circleci\project\build\third_party\ideep\mkl-dnn\src\version.rc
```
Would you please try passing `-DCMAKE_NINJA_CMCLDEPS_RC=0` to CMake?

pinzhenx commented:
@peterjc123 We'll look into that. Thanks for your quick reply.

pinzhenx committed:
dnnl upgrade: update ideep submodule

dzhulgakov reviewed (changes requested):
So there's actually a failing Caffe2 test - AlexNet with MKLDNN rewrite gives wrong results O_O For repro:

`python caffe2/python/mkl/rewrite_graph_test.py MKLRewriteTest.test_mkl_alexnet_rewrite`

I tried to narrow down to the minimal repro: https://gist.github.com/dzhulgakov/a56c19a4cb22acc80edaf56d5cdaaf3f

Looks like FC gives the wrong results if input tensor is not in the canonical layout (the one passed be Conv). I tried to repro similar stuff on PyTorch (using alexnet from torchvision) but simple test seems to have passed.

Can you please investigate further? Maybe C2 integration has some issues.

pinzhenx commented:
@dzhulgakov, thanks a lot for the repro and detailed report. That helps a lot.

The issue actually boils down to the `reshape` which implicitly called by `fc`. It could also be reproduced in pytorch:
```
def test_reshape_blocked_format(self):
    # construct an mkldnn blocked tensor with mkldnn conv2d
    C = 7
    m = mkldnn_utils.to_mkldnn(torch.nn.Conv2d(C, C, 3))
    x = torch.randn(1, C, 8, 8).to_mkldnn()

    # mkldnn tensor w/ blocked format
    y_block = m(x)
    # aten tensor w/ plain format
    y_plain = y_block.to_dense()

    y_block_reshape = y_block.reshape(C, -1)
    y_plain_reshape = y_plain.reshape(C, -1)

    self.assertEqual(y_plain_reshape, y_block_reshape.to_dense())
```

[Patch here](https://github.com/intel/ideep/compare/ada133fb3dfafbe53099f881af35b1883e2e9058...c36993484d09f5ac910631f34504fc3387e5a3bc#diff-999b9f1e883697e8a340ffa32aee1042L683) shows that in the reshape, it unintentionally returned a boolean instead of an int when counting non-one dims. Therefore, it fails to [trigger a reorder](https://github.com/intel/ideep/compare/ada133fb3dfafbe53099f881af35b1883e2e9058...c36993484d09f5ac910631f34504fc3387e5a3bc#diff-999b9f1e883697e8a340ffa32aee1042R692) from blocked format to default plain format as we did before.

pinzhenx committed:
fix reshape issue of blocked tensor

dzhulgakov commented:
Another concern - we're seeing small regression with this PR on single thread batch size 1 inference for some models, e.g. ResNext101. I see in perf comparison you're mostly comparing intra-op parallelism settings (even with batch size 1), but not the single threaded perf.

Specifically for resnet, I've used this benchmark: https://gist.github.com/dzhulgakov/f225e04656458697b99d0adcc6aaed84 with default params (maybe reduce number of iterations) and OMP_NUM_THREADS=1.

It gives about 5% regression on the machine I've tested (260 vs 272ms) with internal build (on D-2191 CPU @ 1.60GHz). I've managed to repro it with CMake build on a different machine (with E5-2650 v4 @ 2.20GHz) and ~3% perf gap.

Would be nice to understand where exactly this gap comes from and eliminate it if possible.

CaoZhongZ commented:
Primitive cache overhead. We switched to DNNL internal primitive cache. It’ll improve as DNNL involve.

Jianhui-Li commented:
> Primitive cache overhead. We switched to DNNL internal primitive cache. It’ll improve as DNNL involve.

Hi @pinzhenx and @CaoZhongZ Could you please provide data on what is the impact of primitive cache overhead, and how widely it impacts the performance negatively? I don't understand why the overhead  doesn't show up in your resnet101 measurement using batch=1 Threads=4, it looks you get perf gain on that setting (116.3%). I usually observed more core we used, the high ratio on overhead.

CaoZhongZ commented:
We are re-testing the same workload on similar hardware see if my theory was right. The test in report generated from skylake server. @Jianhui-Li

pinzhenx commented:
Here's my results from Skylake
```
GCC: gcc version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)
OS: Ubuntu 18.04.4 LTS
CPU: Intel(R) Xeon(R) Platinum 8180 CPU @ 2.50GHz
RAM: 187 G
Config: 50warm 500run 1T
```

ms per iter | new 1.5.0a0+94a4015 | old 1.5.0a0+67608cc
-- | -- | --
resnext50 | 83.85 | 98.32
resnext101 | 149.66 | 159.65
resnext152 | 214.27 | 220.55



<br>

... and this one is from Cascade Lake
```
GCC: gcc version 7.3.1 20180303 (Red Hat 7.3.1-5) (GCC)
OS: CentOS Linux 7 (Core)
CPU: Intel(R) Xeon(R) Platinum 8280L CPU @ 2.70GHz
RAM: 187G
Config: 50warm 500run 1T
```


ms per iter | new 1.5.0a0+94a4015 | old 1.5.0a0+67608cc
-- | -- | --
resnext50 | 78.16 | 93.76
resnext101 | 137.04 | 151.24
resnext152 | 197.08 | 211.38


<br>

However, I was not able to reproduce your performance numbers. From the pytorch profiling results, I observed that the difference was mainly saved in batchnorm, and the conv time were almost the same. 

Could you please provide some other info about your system so that I could find a more similar machine to reproduce it later?

-------

I was comparing this PR with the one you cherrypicked to fix group conv and prepack weights.
```
new:
94a4015e1d6ab8e895dfac606a2ed062d22ed134
fix reshape issue of blocked tensor
 
old:
67608cc018b4548fcccfad2d9969c7d546224736
Fix MKLDNN conv2d 5d weight handling (#34115)
```

dzhulgakov commented:
@pinzhenx - that's bizarre, are you using my benchmarking code, right?

Also, are you running with a single thread? I explicitly pass OMP_NUM_THREADS=1 in my benchmark.

I didn't pass any special build options, just ran plain `setup.py`. For reference here's a build config:

```
$ python -c 'import torch; print(torch.__config__.show())'
PyTorch built with:
  - GCC 7.4
  - C++ Version: 201402
  - Intel(R) MKL-DNN v0.21.1 (Git Hash 7d2fd500bc78936d1d648ca713b901012f470dbc)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - Build settings: BLAS=MKL, BUILD_TYPE=Release, CXX_FLAGS= -fvisibility-inlines-hidden -fopenmp -DNDEBUG -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, FORCE_FALLBACK_CUDA_MPI=1, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=OFF, USE_EIGEN_FOR_BLAS=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=OFF, USE_MKLDNN=ON, USE_MPI=ON, USE_NCCL=OFF, USE_NNPACK=ON, USE_OPENMP=ON, USE_STATIC_DISPATCH=OFF,
```

I just retested resnext101 on CPU E5-2650 v4 @ 2.20GHz using the same commits as you and still see 268us (old) vs 276us (new).

CaoZhongZ commented:
@dzhulgakov Could you share the verbose information with DNNL_VERBOSE=1 or MKLDNN_VERBOSE=1, before and after the patch?

dzhulgakov commented:
@CaoZhongZ - sure, here you go: https://gist.github.com/dzhulgakov/c84cf0c581ddbe3690dea3bc2ec50dae (it's only 20 iterations but the output looks repeatable)

pinzhenx commented:
Thanks @dzhulgakov, we'll take a closer look at it.

pinzhenx commented:
Like @CaoZhongZ mentioned, this gap was indeed introduced by the DNNL cache mechanism as we switched from ideep cache to DNNL built-in cache. We may not fix it immediately and it’ll improve as DNNL involve later.


_Q. What will be affected due to this cache overhead?_

A. All pytorch models may have a 2%+ performance drop, this number may vary from platform to platform, while caffe2 models will not be affected. On the other hand, optimizations like prepacking weights could benefit all pytorch models, giving them a 15%+ improvement, which could compensate for the overhead caused by the cache.



_Q. Why we were not able to reproduce this gap before?_

A. We observed from the log that DNNL is not using avx512 but avx2 instead on your side. For some shapes of kernels in Resnext, we have an optimized path on avx512 for them in this PR, so that's why we were seeing the opposite. Actually, this issue came to light once we changed to an AVX2-only machine.

yinghai commented:
@pinzhenx could you explain a bit more why dnnl caching is degrading the perf by 2%? Is it because that we cache only kernel but not intermediate buffers anymore?

pinzhenx commented:
A quick rundown of DNNL programming model:
```
                                   dnnl cache
                           ___________________________
                          |                           v
params ---------> primitive descriptor ---------> primitive
         create                          create       
    |_________________________________________________^
                      ideep cache
```
To create a `primitive` (kernel), there are two necessary steps:
1. create a `primitive descriptor` (PD) from params like shapes, stride, padding... PD is designed for querying expected tensor formats without fully instantiate a `primitive`
2. create a `primitive` from `primitive descriptor`. This could be a time-consuming process involving  codegen

<br>

ideep cache maintains a map from `params -> primitive`. But how can we query memory formats out from PD if we skip the creation of PD? That's simple, primitive has a pointer to its PD.

However, DNNL cache is built on top of the `PD -> primitive`. As a DNNL user, we need to recreate  PD every time an operation is called.

That's where the overhead comes: **even if the PD creation is highly optimized and expected to be lightweight, it still takes longer than doing nothing.**

The DNNL cache is currently going through a major change, and we may have a pd-cache in addition to primitive cache (or probably other design) in the future to reduce this overhead.


----------------------

>  Is it because that we cache only kernel but not intermediate buffers anymore?

Not quite. Caching buffers brings no benefits for single-batch inference.

CaoZhongZ commented:
Switch to DNNL internal cache is a strategic move. In the future similar primitives could share a lot of stuff when cache is handled by DNNL itself. Should we add old cache scheme back and make a compilation flag?

dzhulgakov commented:
@pytorchbot retest this please

pinzhenx commented:
the current base is kinda buggy. I will rebase it soon.

dzhulgakov commented:
Btw, just to update here on perf results:
* for AVX2 in CMake build there's indeed 1-2% regression which is explained by changing cache behavior. With jemalloc this regression doesn't reproduce, so I've proceeded with landing this PR. Further improvements to cache are very welcome though
* with AVX512 we could reproduce ~10% speed up which matches this PR description
* the regression we saw in internal Buck build even with jemalloc and lack of speed up on avx512 was caused by messed up flags in internal build, now results match those with CMake and jemalloc

Coderx7 commented:
Any update on this?

heraldb commented:
In the mean time, after renaming to DNNL, the library is renamed to oneDNN and is at version 1.5. How about moving directly to oneDNN 1.5?

mingfeima commented:
@heraldb we are working on upgrading to 1.5, while at the same time finish the renaming.
